ToolBox/CoinApi/CoinApiDataQueueHandler.cs

```diff
@@ -0,0 +1,360 @@
+        {
+            if (_subscriptionsPending) return;
+
+            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+            _subscriptionsPending = true;
+
+            Task.Run(() =>
+            {
+                while (true)
+                {
+                    DateTime requestTime;
+                    List<Symbol> symbolsToSubscribe;
+                    lock (_lockerSubscriptions)
+                    {
+                        requestTime = _lastSubscribeRequestUtcTime.Add(_subscribeDelay);
+                        symbolsToSubscribe = _subscribedSymbols.ToList();
+                    }
+
+                    if (DateTime.UtcNow > requestTime)
+                    {
+                        // minimum delay has passed since last subscribe request, send the subscribe request
+                        SubscribeSymbols(symbolsToSubscribe);
+
+                        lock (_lockerSubscriptions)
+                        {
+                            _lastSubscribeRequestUtcTime = DateTime.UtcNow;
+                            if (_subscribedSymbols.Count == symbolsToSubscribe.Count)
+                            {
+                                // no more subscriptions pending, task finished
+                                _subscriptionsPending = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    Thread.Sleep(200);
+                }
+            });
+        }
+
+        /// <summary>
+        /// Returns true if we can subscribe to the specified symbol
+        /// </summary>
+        private static bool CanSubscribe(Symbol symbol)
+        {
+            // ignore unsupported security types
+            if (symbol.ID.SecurityType != SecurityType.Crypto)
+                return false;
+
+            // ignore universe symbols
+            return !symbol.Value.Contains("-UNIVERSE-");
+        }
+
+        /// <summary>
+        /// Subscribes to a list of symbols
+        /// </summary>
+        /// <param name="symbolsToSubscribe">The list of symbols to subscribe</param>
+        private void SubscribeSymbols(List<Symbol> symbolsToSubscribe)
+        {
+            if (symbolsToSubscribe.Count == 0) return;
+
+            Log.Trace($"CoinApiDataQueueHandler.SubscribeSymbols(): {string.Join(",", symbolsToSubscribe)}");
+
+            var message = JsonConvert.SerializeObject(new HelloMessage
+            {
+                ApiKey = _apiKey,
+                Heartbeat = true,
+                SubscribeDataType = new[] { "trade", "quote" },
+                SubscribeFilterSymbolId = _subscribedSymbols.Select(_symbolMapper.GetBrokerageSymbol).ToArray()
+            });
+
+            _webSocket.Send(message);
+
+            _connectionHandler.EnableMonitoring(true);
+        }
+
+        private void OnMessage(object sender, WebSocketMessage e)
+        {
+            //Log.Trace(e.Message);
```

Comment:
This probably was used while testing 😄, should be removed?

Reply:
Sure, will remove 👍